cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+# ISLE: Instruction Selection Lowering Expressions
+
+This document will describe ISLE (Instruction Set Lowering
+Expressions), a DSL (domain-specific language) that we have developed
+in order to help us express certain parts of the Cranelift compiler
+backend more naturally. ISLE was first [described in RFC
+#15](https://github.com/bytecodealliance/rfcs/pull/15) and now is used
+by and lives in the Cranelift tree in
+[cranelift/isle](https://github.com/bytecodealliance/wasmtime/tree/main/cranelift/isle).
+
+## Intro and Whirlwind Tour: DSL for Instruction Lowering
+
+The goal of ISLE is to represent *instruction lowering patterns*. An
+instruction lowering pattern is a specification that a certain
+combination of operators in the IR (CLIF), when combined under certain
+conditions, can be compiled down into a certain sequence of machine
+instructions. For example:
+
+- An `iadd` (integer add) operator can always be lowered to an x86
+  `ADD` instruction with two register sources.
+  
+- An `iadd` operator with one `iconst` (integer-constant) argument can
+  be lowered to an x86 `ADD` instruction with a register and an
+  immediate.
+  
+One could write something like the following in ISLE (simplified from
+the real code [here](../codegen/src/isa/x64/lower.isle):
+
+```lisp
+;; Add two registers.
+(rule (lower (iadd x y))
+      (value_reg (add
+                   (put_in_reg x)
+                   (RegMemImm.Reg (put_in_reg y)))))
+
+;; Add a register and an immediate.
+(rule (lower (iadd x (simm32_from_value y))
+      (value_reg (add
+                   (put_in_reg x)
+                   ;; `y` is a `RegMemImm.Imm`.
+                   y)))
+```
+  
+ISLE lets the compiler backend developer express this information in a
+declarative way -- i.e., just write down a list of patterns, without
+worrying how the compilation process tries them out -- and the ISLE
+DSL compiler will convert this list of patterns into efficient Rust
+code that becomes part of Cranelift.
+
+The rest of this document will describe the semantics of the DSL
+itself. ISLE has been designed to be a general-purpose DSL that can
+apply to any sort of backtracking pattern-matching problem, and will
```

Comment:
We do actually, depending on how you look at it! Specifically one can see the decision-trie traversal as a depth-first attempt at finding a path to a leaf node, where we backtrack up the tree when a match op fails. In the generated Rust this manifests as a control-flow edge that falls out of the `if match_condition { ... }` without returning a `Some(result)`. The "explicit backtracking stack pop" step that one would see in an FSM-like implementation implicitly happens because (i) PC encodes a path through match ops, and we effectively truncate that path by falling through, and (ii) local let-bindings go out of scope, backing up our state.